File: frontend/src/types/game.ts
export type Direction = 'UP' | 'DOWN' | 'LEFT' | 'RIGHT';

export type GameMode = 'walls' | 'pass-through';

export type GameStatus = 'idle' | 'playing' | 'paused' | 'game-over';

export interface Position {
  x: number;
  y: number;
}

export interface GameState {
  snake: Position[];
  food: Position;
  direction: Direction;
  score: number;
  status: GameStatus;
  mode: GameMode;
  speed: number;
}

export interface User {
  id: string;
  username: string;
  email: string;
  createdAt: string;
}

export interface LeaderboardEntry {
  id: string;
  userId: string;
  username: string;
  score: number;
  mode: GameMode;
  createdAt: string;
}

export interface LivePlayer {
  id: string;
  username: string;
  score: number;
  mode: GameMode;
  snake: Position[];
  food: Position;
  direction: Direction;
  isPlaying: boolean;
}

export interface AuthCredentials {
  email: string;
  password: string;
  username?: string;
}

export interface AuthResponse {
  user: User;
  token: string;
}

File: backend/src/schemas/game.py
from datetime import datetime

from pydantic import BaseModel

from .enums import Direction, GameMode


class LeaderboardEntry(BaseModel):
    id: str
    userId: str
    username: str
    score: int
    mode: GameMode
    createdAt: datetime

class ScoreSubmission(BaseModel):
    score: int
    mode: GameMode

class Position(BaseModel):
    x: int
    y: int

class LivePlayer(BaseModel):
    id: str
    username: str
    score: int
    mode: GameMode
    snake: list[Position]
    food: Position
    direction: Direction
    isPlaying: bool
